### ISSUE TITLE
Ifpack2 SingletonFilter does not implement getGraph().

### ISSUE THREAD

**Issue by kddevin** [trilinos/Trilinos#560]
@trilinos/ifpack2 @trilinos/muelu @trilinos/zoltan2 

This issue blocks  #244 , #547, #548.

Ifpack2 uses Zoltan2 for local ordering in Ifpack2_AdditiveSchwarz_def.hpp.  
Zoltan2 works only with scalar_t={double, float, int}.  To prevent compilation problems when scalar_t={std::complex, Sacado type} (issues #244, #547), we would like to use the RowGraph, obtained from RowMatrix::getGraph(), for ordering in Ifpack2.

We have implemented the use of RowGraph in Ifpack2 and Zoltan2 (see #548), and the compilation is fine now.  But some MueLu tests fail with the following thrown error:

 Ifpack2::SingletonFilter: does not support getGraph.

The failing tests are

72:MueLu_Helmholtz2DParallel_MPI_4
73:MueLu_Helmholtz3DParallel_MPI_4
74:MueLu_HelmholtzFEM2DParallel_MPI_4
75:MueLu_HelmholtzFEM3DParallel_MPI_4

@mhoemmen says many of the Filters in Ifpack2 don't correctly implement getGraph().  

I do not know the best way to resolve this problem, as I do not know the purpose of the SingletonFilter, nor how to implement RowGraph for it.  I am open to suggestions from @trilinos/ifpack2 or @trilinos/muelu .

Thank you!

**Comment by mhoemmen** [comment]
@jhux2

**Comment by mhoemmen** [comment]
Sorry for closing temporarily -- hit the wrong button.

**Comment by kddevin** [comment]
@kddevin and @csiefer2  discussed this issue today.  

The conclusion:  getGraph() is difficult to implement in filters without creating a copy of the graph.

The path forward:
Step 1:  For the short term, remove singleton filtering from the MueLu tests so that pull request #548 can be accepted.
Step 2 (Likely in FY17 Q1):  Try reversing the order that singleton filtering and Zoltan2's ordering are applied to the matrix.  Currently, the matrix is filtered and then ordered, requiring the filtered matrix to implement getGraph().  If, instead, we first ordered and then filtered, Zoltan2 could getGraph() from the input RowMatrix, and the filter would not need to implement getGraph().  The resulting matrix should be the same with non-trivial ordering, as the modification from singleton filtering should not affect ordering.

**Comment by egboman** [comment]
I agree with the plan but not sure why "the modification from singleton filtering should not affect ordering."? I think the filtering will change the matrix size and structure so may give different ordering permutation.

**Comment by kddevin** [comment]
OK; I should have been more wishy-washy in my claim, or left this issue to our team's ordering experts instead of taking it on myself.  :) :)
As I understand it, singleton filtering removes rows that have only diagonal entries; thus, a non-trivial ordering like RCM should not have significant differences (since those vertices weren't connected to anything anyway).   Of course, a trivial re-ordering (e.g., random) could have big differences.
In short, we don't expect reversing ordering and singleton filtering will have significant impact on the resulting RCM ordering.

**Comment by egboman** [comment]
I think singleton filtering removes rows with a single nonzero and also the corresponding columns, which may be quite dense (believe this is often the case for circuits). Thus there could be some impact even for RCM.

It is still a reasonable heuristic to try. Note ordering time might increase.

**Comment by csiefer2** [comment]
As per discussions w/ @kddevin , I've disabled Singleton Filtering in those 4 Helmholtz tests.

**Comment by kddevin** [comment]
OK; I misunderstood then.  Can you coordinate with @csiefer2 next quarter to identify/evaluate problems with this approach?  Thanks.

**Comment by egboman** [comment]
OK. Probably will have little or no impact on PDE problems so the MueLu tests should  be fine.

**Comment by maherou** [comment]
I wrote the original singleton filter code many years ago as we were writing solver algorithms for Xyce.  It scans and treats in a special way the equations and variables that are fully independent or fully dependent.  These appear as row (independent) or column (dependent) singletons.

Row singletons are solved and then algebraically removed from the system.  Column singletons are removed and then solved as a post-processing step.  The rest of the equations and variables are solved using any normal Trilinos linear solver.

Often these singletons are part of a very dense column or row, so removing them substantially improves the sparsity of the matrix.  In fact, without removing them, overlapping Schwarz methods are almost useless because of the size of the overlap region is enormous on processors with the singletons.

Singletons are common in circuit problems, but I have not seen them in any other setting.   Passing a PDE based linear system through the singleton filter should have no mathematical impact, unless the Dirichlet boundary conditions are included in the system (row singletons) but will cost time.

History lesson complete ☺

Mike

**Comment by mhoemmen** [comment]
> The conclusion: getGraph() is difficult to implement in filters without creating a copy of the graph.

Not sure if I agree.  getGraph() returns a RowGraph.  This need not be a CrsGraph, so you don't have to copy the graph.  For example, in the case of SingletonFilter, one would just create a "SingletonFilterRowGraph" that keeps the original matrix and filters it, but only offers an interface to the graph.

**Comment by kddevin** [comment]
@hkthorn emailed me the following:
"I've seen the discussion and Mike's history lesson on singleton filtering.  As discussed by Mike and Erik, it needs to be performed before any reordering, as the impact can be huge on circuit matrices.  I've not seen a similar situation in PDE based matrices."

**Comment by mhoemmen** [comment]
I'm calling this a "task," because the work to do is well defined (implement SingletonRowGraph), and could be done by one person.

**Comment by brian-kelley** [comment]
Sorry, my push just now was for #460, not this.

**Comment by brian-kelley** [comment]
Sorry @bmpersc I put the wrong issue number in my commit message - I actually fixed #460.
